src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/ContractsController.kt

```diff
@@ -0,0 +1,185 @@
+
+/**
+ * The controller for contracts
+ */
+@ApplicationScoped
+class ContractsController: AbstractSapResourceController() {
+    @Inject
+    private lateinit var sapSessionController: SapSessionController
+
+    @Inject
+    private lateinit var logger: Logger
+
+    @Inject
+    private lateinit var configController: ConfigController
+
+    /**
+     * Lists contracts
+     *
+     * @param startDate start date filter
+     * @param businessPartnerCode business partner code filter
+     * @param contractStatus contract status filter
+     * @return contracts
+     */
+    fun listContracts(startDate: LocalDate?, businessPartnerCode: String?, contractStatus: SapContractStatus?): List<JsonNode> {
+        sapSessionController.createSapSession().use { sapSession ->  
+            val resourceUrl = "${sapSession.apiUrl}/BlanketAgreements"
+            val startDateFilter = startDate?.let { "StartDate ge '$startDate'" }
+            val businessPartnerCodeFilter = businessPartnerCode?.let { "BPCode eq '$businessPartnerCode'" }
+            val contractStatusFilter = contractStatus?.let { "Status eq '${contractStatusToSapFormat(contractStatus)}'" }
+            val combinedFilter = listOfNotNull(startDateFilter, businessPartnerCodeFilter, contractStatusFilter).joinToString(" and ")
+
+            if (combinedFilter.isEmpty()) {
+                val contracts = getItemsAsJsonNodes(
+                    resourceUrl = resourceUrl,
+                    select = "\$select=*",
+                    routeId = sapSession.routeId,
+                    sessionId = sapSession.sessionId
+                )
+                return compressContracts(contracts = contracts, session = sapSession)
+            }
+
+            val filter = "\$filter=$combinedFilter"
+
+            val contracts = getItemsAsJsonNodes(
+                resourceUrl = resourceUrl,
+                select = "\$select=*",
+                filter = filter,
+                routeId = sapSession.routeId,
+                sessionId = sapSession.sessionId
+            )
+            return compressContracts(contracts = contracts, session = sapSession)
+        }
+    }
+
+    /**
+     * Compresses contracts. From one contract per item to one contract per item group.
+     *
+     * @param contracts contracts to compress
+     * @param session SAP session to use
+     * @return compressed contracts
+     */
+    private fun compressContracts(contracts: List<JsonNode>, session: SapSession): List<JsonNode> {
+        val newContracts = mutableListOf<JsonNode>()
+        val groupCodes = configController.getGroupCodesFile()
+        val itemPropertiesSelect = getItemPropertiesSelect(groupCodes = groupCodes)
+        val items = getItemsAsJsonNodes(
+            resourceUrl = "${session.apiUrl}/Items",
+            select = itemPropertiesSelect,
+            sessionId = session.sessionId,
+            routeId = session.routeId
+        )
+
+        contracts.forEach { contract ->
```

Comment:
Split multiple for each loops to own funtions

Reply:
Splitted